src/main/java/com/coveo/spillway/storage/RedisStorage.java

```diff
@@ -55,57 +58,144 @@
 public class RedisStorage implements LimitUsageStorage {
 
   private static final String KEY_SEPARATOR = "|";
+  private static final String DEFAULT_PREFIX = "spillway";
 
-  private final Jedis jedis;
+  private final JedisPool jedisPool;
   private final String keyPrefix;
 
+  public RedisStorage(URI uri) {
+    this(uri, DEFAULT_PREFIX);
+  }
+
+  public RedisStorage(URI uri, String prefix) {
+    this(uri, prefix, Protocol.DEFAULT_TIMEOUT);
+  }
+
+  public RedisStorage(URI uri, String prefix, int timeout) {
+    this(new GenericObjectPoolConfig(), uri, prefix, timeout);
+  }
+
+  public RedisStorage(GenericObjectPoolConfig poolConfig, URI uri) {
+    this(poolConfig, uri, DEFAULT_PREFIX);
+  }
+
+  public RedisStorage(GenericObjectPoolConfig poolConfig, URI uri, String prefix) {
+    this(poolConfig, uri, prefix, Protocol.DEFAULT_TIMEOUT);
+  }
+
+  public RedisStorage(GenericObjectPoolConfig poolConfig, URI uri, String prefix, int timeout) {
+    jedisPool = new JedisPool(poolConfig, uri, timeout);
+    keyPrefix = prefix;
+  }
+
   public RedisStorage(String host) {
-    this(new Jedis(host));
+    this(host, Protocol.DEFAULT_PORT);
   }
 
   public RedisStorage(String host, int port) {
-    this(new Jedis(host, port));
+    this(host, port, DEFAULT_PREFIX);
   }
 
-  public RedisStorage(URI uri) {
-    this(new Jedis(uri));
+  public RedisStorage(String host, int port, String prefix) {
+    this(new GenericObjectPoolConfig(), host, port, prefix);
+  }
+
+  public RedisStorage(GenericObjectPoolConfig poolConfig, String host) {
+    this(poolConfig, host, Protocol.DEFAULT_PORT);
   }
 
+  public RedisStorage(GenericObjectPoolConfig poolConfig, String host, int port) {
+    this(poolConfig, host, port, DEFAULT_PREFIX);
+  }
+
+  public RedisStorage(GenericObjectPoolConfig poolConfig, String host, int port, String prefix) {
+    this(poolConfig, host, port, prefix, Protocol.DEFAULT_TIMEOUT);
+  }
+
+  public RedisStorage(
+      GenericObjectPoolConfig poolConfig, String host, int port, String prefix, int timeout) {
+    jedisPool = new JedisPool(poolConfig, host, port, timeout);
+    keyPrefix = prefix;
+  }
+
+  public RedisStorage(
+      GenericObjectPoolConfig poolConfig,
+      String host,
+      int port,
+      String prefix,
+      int timeout,
+      String password) {
+    this(poolConfig, host, port, prefix, timeout, password, Protocol.DEFAULT_DATABASE);
+  }
+
+  public RedisStorage(
+      GenericObjectPoolConfig poolConfig,
+      String host,
+      int port,
+      String prefix,
+      int timeout,
+      String password,
+      int database) {
+    jedisPool = new JedisPool(poolConfig, host, port, timeout, password, database);
+    keyPrefix = prefix;
+  }
+
+  /**
+   * Will be remove in V1.2.0
+   *
+   * @param jedis A Jedis connection
+   */
+  @Deprecated
   public RedisStorage(Jedis jedis) {
```

Comment:
Too bad, it's a basic concept in dependency injection and very useful to ensure maximum flexibility.

For example if a client needs to set something that we haven't exposed, we don't have to add a new constructor. We can simply accept the JedisPool directly. e.g. if someone needs to set something like proxy settings or whatever setting that we do not care about.